_posts/2019-01-30-year-of-the-dog.md

```diff
@@ -0,0 +1,199 @@
+
+Faced with these problems, we decided to try and address them by progressively
+introducing a new GraphQL API, in the same Rails codebase as the existing
+RESTful API. Here is what we observed:
+
+1. The additional layer gave us more room and structure to **move business
+   logic back to the backend**.
+2. **Data fetching in the frontend is now a lot less complex**, with queries
+   mapping more or less 1-to-1 with views, and mutations to forms and buttons.
+   We are not using any redux-based abstraction for CRUD anymore, and the
+   overall amount of state management code has decreased. State is (or
+   appears) also more localized to where it belongs in the component tree. We
+   found this easier to reason about.
+3. The simplicity, but also the adherence to a well-adopted and documented
+   standard for data fetching made **onboarding easier**. There are a lot of
+   good GraphQL resources for frontend developers. In our experience picking up
+   the technology was easy for everyone.
+4. The tooling is amazing, it lets us **leverage Typescript's strengths by
+   generating precise types for our API data**. We use apollo-client and
+   apollo-codegen, they are well-documented, worked fine and the caching layer
+   is the cherry on top.
+5. **Our API documentation is now in much better shape**, because most of it is
+   automated, keeping it in sync with the implementation is trivial, and
+   [GraphiQL][graphiql] offers a good interface to explore it.
+
+It took some time to have a viable minimum implementation, but once this was
+deployed, it became clear very quickly that we wanted to go in that direction.
+
+Two more pain-points that we did not identify before migrating but are now obvious:
+
+- **Debuggability**. [GraphiQL][graphiql] has been a boon to identify where
+  errors come from.  With multiple requests that cannot be reproduced easily,
+  it was sometimes hard to distinguish between backend and frontend issues (the
+  network tools do not let you tweak and reproduce the requests easily). With
+  queries you can share directly as a link — [GraphiQL][graphiql] can take
+  queries from URL query parameters — we saved a lot of debugging time.
+- **Evolvability**. The way fields can be added and deprecated progressively
+  has radically decreased the frequency of necessary breaking changes and made
+  deployment smoother as a result. We use
+  [eslint-plugin-graphql](https://github.com/apollographql/eslint-plugin-graphql)
+  to warn us when an app is using a field that has been deprecated.
+
+![store2be developer discovering GraphiQL]({{ site.url }}/images/2018_retrospective/developer_discovering_graphql.jpg)
+_store2be developer discovering GraphiQL (photo © unknown)_
+
+## GraphQL in our Rails API
+
+Adopting GraphQL was in retrospect an easy win in the frontend. It did require
+some changes, but towards a more standard workflow with a wide array of tools
+to make our lives easier. It was not so in the API. GraphQL APIs pose a
```

Comment:
> it was not so ... in the API.

Feels like it is missing an adjective, y'know.

Reply:
Replaced with "it was not the same in the API"